### DISCUSSION TITLE
Missing block design of marga_system

### DISCUSSION THREAD

**Discussion by volatile-static** [vnegnev/marcos_extras#13] [Q&A] [answered]
I am trying to port marga_system from `rp-122` to our own Zynq board. I would like to ask you some questions and hope you can spare some time to reply.

First of all, I would like to know how to generate the new version of the FPGA project, because I did not find the Makefile in the marga_system repository, while the flocra_system provides the Makefile and related tcl. I am not sure if I need to copy the Makefile myself or if there is another way to build the project.

Secondly, I would like to know which source file contains the newly added [receive enable](https://github.com/vnegnev/marga/commit/aec5026b423b023889d4e49a8d31c154c23390fb#diff-9c31404880fc4f7e29997a178b71ba3debe8959b6c81df326b7d996c4cf49592R236) port, because I saw this port setting in [flomachine.py](https://github.com/vnegnev/marcos_client/commit/ad875244de6bebb9824d95d56e7b05e91a4f03ef#diff-a9c84e5d2e168ca2afb9ad6397c9e71c40344edbc032fd1a306932ff542ac98bR32), but I did not find the corresponding `block_design.tcl` code. I would like to know the function and implementation of this port.

Thank you very much for your contribution and help, looking forward to your reply.

Best regards

**Comment by vnegnev** [comment, marked as answer]
Hi, sorry about the delay! I seem to have problems getting notifications...

The FPGA firmware is separate from marga_system at the moment. You can currently find it here: https://github.com/vnegnev/marcos_fpga/tree/rename-marga . I had to fix a few bugs, but now the `rename-marga` branch works with Vivado + Vitis 2020.2 - please let me know if the instructions in the readme are sufficient. Note that this branch uses @catkira's open-source DDS, complex multiplier and CIC cores, rather than the Xilinx cores used in earlier branches.

Regarding the receive enable port, I'm afraid I don't fully understand your question. The receive enable output from the marga core was previously used to reset the RX chain (i.e. external to the marga core) when a receive window wasn't being run, but IIRC this caused problems in the Xilinx CIC core (which had a really long latency after reset before it started running correctly) so instead I used it to just pause the AXI stream of I/Q samples from the DDS to the RX chain. To see the difference more clearly, you may want to rebuild the block design from an earlier commit using the Xilinx cores to compare; the open-source cores behave a bit differently from the original Xilinx cores and I want to add more testbenches to marga that simulate the whole RX/TX chain (i.e. marga + the open-source cores). Does that answer your question?

![image](https://github.com/vnegnev/marcos_extras/assets/4116364/b118a380-025e-4ad2-aa18-8e2cfb2443b5)

![image](https://github.com/vnegnev/marcos_extras/assets/4116364/6ed32c17-031e-4866-b61f-85cc1646141e)

**Reply by volatile-static** [reply]
Hi, vnegnev
Thanks very much for your reply!! I run make and got some errors like this：
```shell
ERROR: [IP_Flow 19-627] Port 's_axis_a_tdata': XPath expression failed: Undefined parameter "EFF_PORT_WIDTH_A" used in expression "(EFF_PORT_WIDTH_A - 1)".
ERROR: [IP_Flow 19-627] Port 's_axis_b_tdata': XPath expression failed: Undefined parameter "EFF_PORT_WIDTH_B" used in expression "(EFF_PORT_WIDTH_B - 1)".
ERROR: [IP_Flow 19-627] Port 'm_axis_dout_tdata': XPath expression failed: Undefined parameter "EFF_PORT_WIDTH_OUT" used in expression "(EFF_PORT_WIDTH_OUT - 1)".
CRITICAL WARNING: [IP_Flow 19-3834] Port 'm_axis_dout_tdata': Port mapped to the logical port 'TDATA' of 'xilinx.com:interface:axis:1.0' interface must have a width of multiples of 8 bits.  The current port width is '1'.
INFO: [IP_Flow 19-2181] Payment Required is not set for this core.
INFO: [IP_Flow 19-2187] The Product Guide file is missing.
ERROR: [Common 17-39] 'ipx::package_project' failed due to earlier errors.

    while executing
"ipx::package_project -import_files -root_dir tmp/cores/$core_name"
    (file "scripts/core.tcl" line 24)
INFO: [Common 17-206] Exiting Vivado at Wed Jul 19 11:17:31 2023...
make: *** [Makefile:53: tmp/cores/complex_multiplier_v1_0] Error 1
```
In addition, I have to copy the sv files from submodules to cores manually after clone, otherwise there were only symbolic links.
By the way, I use `make -j` to enable maximum CPUs.

regards,

**Reply by vnegnev** [reply]
Sorry, @catkira has found the problem - instead of running `git submodule update --init --remote`, just run `git submodule update --init` . I've updated the readme with the correct command - just tried it in a clean clone and the Makefile ran, let me know if you have any problems!

> In addition, I have to copy the sv files from submodules to cores manually after clone, otherwise there were only symbolic links.

If the `git submodule` command runs correctly, it'll clone the repos into the folders that the symlinks link to - you shouldn't have to copy anything manually.

>  By the way, I use make -j to enable maximum CPUs.

Yep, also an option but Vivado has locked up for me when doing this before - with 4 cores it seemed to work fine. I haven't seen the error in a few years though so maybe it's fine with `make -j` as well!

**Reply by vnegnev** [reply]
What kind of Zynq board are you porting it to? I'm always interested in other options than the rp-122!

**Comment by volatile-static** [comment]
Hi, @catkira & @vnegnev 
I've tried `git submodule update --init` and got a same result.
![image](https://github.com/vnegnev/marcos_extras/assets/59945270/1e4a195e-b3c3-48fb-acd1-9f61364912bb)
```shell
 git clone git@github.com:vnegnev/marcos_fpga.git --recursive --single-branch --branch rename-marga
```
All the same.

> What kind of Zynq board are you porting it to? I'm always interested in other options than the rp-122!

We use a `XC7Z100-2FFG900` with 250MHz clock and the marga seems work well.
![image](https://github.com/vnegnev/marcos_extras/assets/59945270/08b2674b-08ea-4a62-9d40-9d6276fdd029)

**Reply by catkira** [reply]
Hi,
just make sure the complex_multiplier submodule is on commit e22f0d139dc9b30b9ec4a9deb2d3c98a5682b613 . That should work.

**Reply by volatile-static** [reply]
But in [that branch](https://github.com/vnegnev/marcos_fpga/tree/rename-marga/hdl/submodules), it's not e22f0d1:
![image](https://github.com/vnegnev/marcos_extras/assets/59945270/8b920364-b271-4bb1-98c3-18b79ea79e11)
Maybe you should update the link.

**Reply by catkira** [reply]
yes, until then just change it manually with git reset or git checkout.

**Reply by volatile-static** [reply]
I copied e22f0d1 manually and build successfully, thank you!
